File: src/html_tools/heatmap.js
(function() {

    const heatmap_data___id__ = __data__;
    const min_cells_x = __min_cells_x__;

    let heatmap_filters = __filters__;
    let render_timeout = null;

    for (const dim of "xy") {
        let elem = document.querySelector(".heatmap-filters-__id__ input.filter-" + dim);
        if (elem) {
            elem.addEventListener("input", function (e) {
                heatmap_filters[dim] = e.target.value;
                render_heatmap_lazy();
            });
        }
    }

    document.addEventListener("load", function () {
        render_heatmap();
    });

    function render_heatmap_lazy() {
        if (render_timeout)
            window.clearTimeout(render_timeout);

        render_timeout = window.setTimeout(render_heatmap, 250);
    }

    function filter_labels_to_index(labels, filter, is_x) {
        let index = [];
        for (let x=0; x < labels.length; ++x)
            index.push(x);

        if (filter && filter.length) {
            const filters = filter.split(",").map(function(f) { return new RegExp(f.trim()); });
            index = index.filter(function(i) {
                for (const f of filters) {
                    if (labels[i].match(f))
                        return true;
                }
            });
        }

        if (is_x) {
            if (index.length < min_cells_x)
                index.push(labels.length);
            /*for (let i=index.length; i < min_cells_x; ++i)
                index.push(i + labels.length);*/
        }

        return index;
    }

    function render_heatmap() {
        const data = heatmap_data___id__;
        const num_colors = __num_colors__;

        let index_x = filter_labels_to_index(data.labels_x, heatmap_filters.x, true),
            index_y = filter_labels_to_index(data.labels_y, heatmap_filters.y);

        let min_val = 0., max_val = 0.;
        for (const y of index_y) {
            const row = data.matrix[y];
            for (const x of index_x) {
                let value = row[x];
                if (typeof value === "number" && !isNaN(value)) {
                    min_val = Math.min(min_val, value);
                    max_val = Math.max(max_val, value);
                }
            }
        }
        const color_factor = min_val === max_val ? 0. : (num_colors) / (max_val - min_val);

        let html = "";

        function render_x_labels(class_name) {
            html += `<div class="${class_name}"></div>`;
            for (const x of index_x) {
                const label = data.labels_x[x];
                if (label)
                    html += `<div class="${class_name}" title="${label}">${label}</div>`;
                else
                    html += `<div class="${class_name}"></div>`;
            }
        }
        render_x_labels("hmlabelv");

        for (const y of index_y) {
            const row = data.matrix[y];

            const label = data.labels_y[y];
            html += `<div class="hmlabel" title="${label}">${label}</div>`;

            for (const x of index_x) {
                if (x < data.labels_x.length) {
                    const
                        value = row[x],
                        title = `${data.labels_x[x]} / ${data.labels_y[y]}: ${value}`;

                    let col = "empty";
                    if (typeof value === "number" && !isNaN(value))
                        col = Math.min(__num_colors__ - 1, Math.floor((value - min_val) * color_factor));

                    html += `<div class="hmc hmc-${col}" title="${title}"></div>`;
                }
                else {
                    html += `<div class="hmc hmc-overlap"></div>`;
                }
            }
        }

        render_x_labels("hmlabelvb");

        const elem = document.querySelector(".heatmap-__id__");
        const need_extra_space = (index_x[index_x.length-1] >= data.labels_x.length);
        const num_cells = need_extra_space ? index_x.length - 1 : index_x.length;
        let columns = `__label_width__ repeat(${num_cells}, 1fr)`;
        if (need_extra_space)
            columns += ` ${min_cells_x - index_x.length + 1}fr`;
        elem.style = `grid-template-columns: ${columns}`;
        elem.innerHTML = html;
    }

    render_heatmap();

})();

File: src/html_tools/heatmap.py
import pathlib
import json
import secrets
import html as html_lib
from typing import Union, Optional, List

import pandas as pd


_JS_SOURCE = pathlib.Path(__file__).resolve().parent.joinpath("heatmap.js").read_text()

assert "//" not in _JS_SOURCE, f"Replace // ... with /* ... */ to make it one-line-able."


def html_heatmap(
        matrix: Union[List[List], pd.DataFrame],
        labels_x: Optional[List[str]] = None,
        labels_y: Optional[List[str]] = None,
        title: str = None,
        transpose: bool = False,
        filterable: bool = True,
        filter_x: str = None,
        filter_y: str = None,
        show_empty_x: bool = True,
        show_empty_y: bool = True,
        display: bool = True,
        colors: Union[str, List[str]] = "GnBu",
        label_width: str = "10rem",
        max_label_length: int = 30,
        keep_label_front: bool = False,
        min_cells_x: int = 20,
        max_cells_x: int = 100,
        max_cells_y: int = 100,
):
    if display:
        from IPython.display import display, HTML

    if not isinstance(matrix, pd.DataFrame):
        matrix = pd.DataFrame(matrix)

    if not labels_x:
        labels_x = matrix.columns
    if not labels_y:
        labels_y = matrix.index

    has_empty_cells = matrix.isnull().values.any()

    if transpose:
        matrix = matrix.transpose()
        labels_x, labels_y = labels_y, labels_x

    if isinstance(colors, str):
        color_name = colors
        try:
            from .colors import get_color_map
        except ImportError:
            from colors import get_color_map
        colors = get_color_map(color_name)

        if not colors:
            raise ValueError(f"No color map found with name '{color_name}'")

    ID = secrets.token_hex(10)

    html = """<style>
        .heatmap-%(ID)s {
            display: grid;
        }
        .heatmap-filters {
            color: #a0a0a0;
        }
        .heatmap-filters-axis {
            display: grid;
        }
        .heatmap-filters-axis label {
            display: inline;
        }
        .heatmap-filters-axis div {
            grid-row: row;
        }
        .heatmap-filters-axis .heatmap-filter-cell-string {
            grid-column: 1 / 6;
        }
        .heatmap-filters-axis .heatmap-filter-cell-page {
            width: 14rem;
        }
        .heatmap-filters-axis input[type="text"] {
            width: 85%%;
        }
        .heatmap-filters-axis input[type="number"] {
            width: 4rem;
        }
        
        .heatmap-grid {
            display: grid;
        }
        .heatmap-%(ID)s .hmlabel,
        .heatmap-%(ID)s .hmlabelv,
        .heatmap-%(ID)s .hmlabelvb {
            text-align: right;
            padding-right: .3rem;
            /*overflow: hidden;*/
            white-space: nowrap;
        }
        .heatmap-%(ID)s .hmlabelv,
        .heatmap-%(ID)s .hmlabelvb {
            writing-mode: sideways-lr;
            /*height: %(label_width)s;*/
            padding: 0;
            padding-bottom: .3rem;
            text-align: left;
            width: 0;
            line-height: 0.6rem;
        }
        .heatmap-%(ID)s .hmlabelvb {
            text-align: right;
            padding: 0;
            padding-top: .3rem;
        }
        .heatmap-%(ID)s .hmc {
            padding-bottom: 100%%;
            border-top: 1px solid rgba(255, 255, 255, .2);
            border-left: 1px solid rgba(255, 255, 255, .1);
            border-right: 1px solid rgba(0, 0, 0, .1);
            border-bottom: 1px solid rgba(0, 0, 0, .3);
        }
        .heatmap-%(ID)s .hmc.hmc-empty {
            border: none;
            border-left: 1px solid rgba(0, 0, 0, .2);
            border-top: 1px solid rgba(0, 0, 0, .2);
            border-right: 1px solid rgba(255, 255, 255, .2);
            border-bottom: 1px solid rgba(255, 255, 255, .2);
        }
        .heatmap-%(ID)s .hmc-overlap {
            border: none;
            padding: 0;
        }
    """ % {
        "ID": ID,
        "width": matrix.shape[1],
        "label_width": label_width,
    }
    for i, c in enumerate(colors):
        html += f""".heatmap-{ID} .hmc.hmc-{i} {{ background: {c} }}"""
    html += """</style>"""

    if title:
        html += f"<h3>{title}</h3>"

    html += f"""<div class="heatmap-{ID}">"""

    if filterable:
        def render_axis_params(dim: str, filter_value: str, show_empty: bool, page: int) -> str:
            return f"""
                <div class="heatmap-filters-axis heatmap-filters-axis-{dim}">
                    <div class="heatmap-filter-cell-string"><label>
                        {dim} filter <input type="text" value="{filter_value or ""}"></label>
                    </div>
                    <div class="heatmap-filter-cell-empty" {"hidden" if not has_empty_cells else ""}><label>
                        show empty <input type="checkbox" {"checked" if show_empty else ""}></label>
                    </div>
                    <div class="heatmap-filter-cell-page"><label>
                        page <input type="number" value="{page+1}" min="1"> of <span class="heatmap-page-count">1</span></label>
                    </div>
                </div>
            """
        html += f"""
            <div class="heatmap-filters">
                {render_axis_params("x", filter_x, show_empty_x, 0)}
                {render_axis_params("y", filter_y, show_empty_y, 0)}
                <label class="heatmap-dimensions"></label>    
            </div>
        """

    html += f"""<div class="heatmap-grid"></div></div>"""

    script_context = {
        "id": ID,
        "num_colors": len(colors),
        "label_width": label_width,
        "filters": json.dumps({
            "x": filter_x,
            "y": filter_y,
            "empty_x": show_empty_x,
            "empty_y": show_empty_y,
            "page_x": 0,
            "page_y": 0,
            "offset_x": 0,
            "offset_y": 0,
        }),
        "max_label_length": max_label_length,
        "keep_label_front": json.dumps(keep_label_front),
        "min_cells_x": min_cells_x,
        "max_cells_x": max_cells_x,
        "max_cells_y": max_cells_y,
        "data": json.dumps({
            "matrix": matrix.values.tolist(),
            "labels_x": [html_lib.escape(str(l)) for l in labels_x],
            "labels_y": [html_lib.escape(str(l)) for l in labels_y],
        }),
    }
    script = _JS_SOURCE
    for key, value in script_context.items():
        script = script.replace(f"__{key}__", str(value))

    html += f"""<script type="text/javascript">{script}</script>"""

    html = " ".join(line.strip() for line in html.splitlines())

    if display:
        display(HTML(html))
    else:
        return html


if __name__ == "__main__":
    import random
    import numpy as np

    if 1:
        def random_name():
            return "".join(
                random.choice(["ku", "ka", "bo", "ba", "su", "la", "to", "mi", "no", "ha", '\"', ">", "<"])
                for _ in range(random.randrange(3, 15))
            )

        rows = []
        value = 0
        for y in range(1000):
            row = []
            for x in range(200):
                row.append(value if random.randint(0, 1) else np.nan)
                value += 1
            rows.append(row)

        labels_x = [f"{i}-{random_name()}" for i in range(len(rows[0]))]
        labels_y = [f"{i}-{random_name()}" for i in range(len(rows))]

        html = html_heatmap(
            rows,
            labels_x=labels_x,
            labels_y=labels_y,
            display=False,
            label_width="15rem",
            max_label_length=20,
            keep_label_front=True,
            min_cells_x=40,
            #max_cells_y=100,
            #colors=["red", "green", "blue"],
        )
    else:
        df = pd.read_pickle("test-data.pkl")
        html = html_heatmap(
            df,
            transpose=True,
            display=False,
            filter_x="m",
        )

    html = f"""
    <!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
    </head>
    <body>
        <h3>heatmap test</h3>
        {html}
    </body>
    </html>
    """
    with open("heatmap-test.html", "w") as fp:
        fp.write(html)
